app/guided.py

```diff
@@ -0,0 +1,413 @@
+        for supporting_link in supporting_links:
+            await self.operation.apply(supporting_link)
+        await self.operation.apply(link_to_execute)
+        self.last_action = link_to_execute.ability
+        return link_to_execute
+
+    async def _get_supporting_links(self, link, abilities):
+        """
+        This method is a quick fix that ideally should be resolved by improvements to abilities at some point in the future. 
+        It's here to capture orphaned abilities that do not have any output facts. Within the dependency graph, they would
+        look like single nodes that are not directly connected to any other components. It is still possible given the way
+        that some abilities are implemented that some of these orphaned abilities are expected to be run before others even
+        through there is not a fact dependency between them. This method gets around this by selecting all links with no
+        output facts that are earlier in the atomic ordering that the currently chosen link. This guarantees that all
+        possible dependent actions are taken prior to the chosen action.
+        """
+        async def has_output_facts(ability, agent):
+            executor = await agent.get_preferred_executor(ability)
+            if not executor:
+                return False
+            for parser in executor.parsers:
+                if parser.parserconfigs:
+                    return True
+            return False
+        
+        def get_list_index(item, list_):
+            if item not in list_:
+                return -1
+            for i, value in enumerate(list_):
+                if value == item:
+                    return i
+
+        if self.last_action:
+            last_ability_index = get_list_index(self.last_action.ability_id, abilities)
+        else:
+            last_ability_index = 0
+        link_ability_index = get_list_index(link.ability.ability_id, abilities)
+
+        agent = (await self.data_svc.locate('agents', match=dict(paw=link.paw)))[0]
+        links = await self.planning_svc.get_links(self.operation, agent=agent)
+        connecting_abilities = abilities[last_ability_index:link_ability_index]
+        connecting_abilities = await self.data_svc.locate('abilities', match=dict(ability_id=tuple(connecting_abilities)))
+        connecting_abilities = [ability for ability in connecting_abilities if not await has_output_facts(ability, agent)]
+        supporting_links = [l for ability in connecting_abilities for l in links if l.ability == ability]
+
+        return supporting_links
+
+    async def _update_goals(self, current_goals):
+        """
+        Removes goals from the set that have been successfully accomplished.
+        """
+        remaining_goals = []
+        for goal in current_goals:
+            goal_found = False
+            if goal.satisfied(await self.operation.all_facts()):
```

Comment:
Is this an expensive operation? 

Should we cache the knowledge that we know previous goals have been met?

Reply:
The output of this method replaces the set of goals, so once a goal has been satisfied it is no longer in the running to be checked again. So it functionally does that already.